StackTracerCore/Debugger.cs

```diff
@@ -46,175 +47,168 @@ private static void TraceWarning(string message, Exception ex)
+                                };
+
+                                var stackTrace = new List<string>();
+
+                                var threadStackTrace = thread.EnumerateStackTrace();
+                                TraceLine($"EnumerateStackTrace returned = {threadStackTrace.Count()} frames", false);
+
+                                foreach (ClrStackFrame frame in threadStackTrace)
                                 {
-                                    var clrMethod = frame.Method;
-                                    if (clrMethod != null && clrMethod.Signature != null)
+                                    if (frame == null)
                                     {
-                                        stackTrace.Add(clrMethod.Signature);
+                                        continue;
                                     }
-                                }
-                                else
-                                {
-                                    var stackFrame = frame.FrameName;
-                                    if (!string.IsNullOrEmpty(stackFrame))
+
+                                    if (frame.Kind == ClrStackFrameKind.ManagedMethod)
+                                    {
+                                        var clrMethod = frame.Method;
+                                        if (clrMethod != null && clrMethod.Signature != null)
+                                        {
+                                            stackTrace.Add(clrMethod.Signature);
+                                        }
+                                    }
+                                    else
                                     {
-                                        stackTrace.Add(stackFrame);
+                                        var stackFrame = frame.FrameName;
+                                        if (!string.IsNullOrEmpty(stackFrame))
+                                        {
+                                            stackTrace.Add(stackFrame);
+                                        }
                                     }
                                 }
-                            }
 
-                            if (stackTrace.Count > 3)
-                            {
-                                DateTime dtTemp = DateTime.UtcNow;
-                                t.CallStackHash = string.Join(Environment.NewLine, stackTrace).GetHashCode();
-                                if (!StackTraceHashes.Contains(t.CallStackHash))
+                                if (stackTrace.Count > 3)
                                 {
-                                    StackTraceHashes.Add(t.CallStackHash);
-                                    t.CallStack = stackTrace;
+                                    t.CallStackHash = string.Join(Environment.NewLine, stackTrace).GetHashCode();
+                                    if (!StackTraceHashes.Contains(t.CallStackHash))
```

Comment:
StackTraceHashes is defined as `static readonly List<int> StackTraceHashes = new List<int>();`.  It would be better if `StackTraceHashes` were a HashSet<int>.  Then you could use:

``` C#
if (StackTraceHashes.Add(t.CallStackHash))
{
    t.CallStack = stackTrace;
}
stacks.Add(t);
```

Using a List<int> incurs a linear search for the hash.  This might be pretty slow if there's a lot of threads.

Reply:
Thanks for pointing this out. Changed.